libs/constrained-generators/src/Constrained/GenT.hs

```diff
@@ -31,24 +31,79 @@ module Constrained.GenT where
 
 import Control.Monad
 import Data.Foldable
+import Data.List.NonEmpty (NonEmpty ((:|)), (<|))
 import Data.List.NonEmpty qualified as NE
 import GHC.Stack
 import System.Random
 import Test.QuickCheck hiding (Args, Fun)
 import Test.QuickCheck.Gen
-import Test.QuickCheck.Random
+
+-- ==============================================================
+-- The GE Monad
+
+-- | It distinguishes between two kinds of errors: FatalError and GenError
+-- and non-fatal errors.
+data GE a
+  = FatalError (NonEmpty (NonEmpty String))
+  | GenError (NonEmpty (NonEmpty String))
+  | Result a
+  deriving (Ord, Eq, Show, Functor)
+
+instance Applicative GE where
+  pure = Result
+  (<*>) = ap
+
+instance Monad GE where
+  FatalError es >>= _ = FatalError es
+  GenError es >>= _ = GenError es
+  Result a >>= k = k a
 
 ------------------------------------------------------------------------
--- The Gen Error monad
+-- The GenT monad
+-- An environment monad on top of GE
 ------------------------------------------------------------------------
 
+-- | Generation mode - how strict are we about requiring the generator to
+-- succeed. This is necessary because sometimes failing to find a value means
+-- there is an actual problem (a generator _should_ be satisfiable but for
+-- whatever buggy reason it isn't) and sometimes failing to find a value just
+-- means there are no values. The latter case is very relevant when you're
+-- generating e.g. lists or sets of values that can be empty.
+data GenMode
+  = Loose
+  | Strict
+  deriving (Ord, Eq, Show)
+
+newtype GenT m a = GenT {runGenT :: GenMode -> [NonEmpty String] -> Gen (m a)}
```

Comment:
To be honest with you I don't fully understand how addition of this non empty list of message helps us.

Unrelated note that I wanted to point it out that if we were to use `quickcheck-transformers` then this could have been changed to using [`GenT`](https://hackage.haskell.org/package/quickcheck-transformer-0.3.1.2/docs/Test-QuickCheck-GenT.html#t:GenT).
I am not suggesting we do it right now, but here is how it would look like:

```haskell
import qualified Test.QuickCheck.GenT as QCT

data Env = 
  Env { envMode :: GenMode
      , envMessages :: [NonEmpty String]
      }

newtype GenT m a = GenT {runGenT :: QCT.GenT (ReaderT Env m) a}
```

This would allow us to derive a lot of instances and use the abstract interface provided by that package.

Reply:
the [NonEmpty String]  accumulates explanations when ever one uses explain. If at some point you call (fatalError xs) xs is added to the accumulated messages and the system fails. If you never call fail, they are thrown away.  The old system added explanations, but they never got thrown away. So on an error, there would be explanations about errors that never happened.

Reply:
Ohh yeah, I see. That totally makes sense now. `explain` works similar to `local` for the `ReaderT`, in a sense that it will annotate the action supplied as an argument, but once the action is done without a failure any added annotation is discarded.

Very nice!